test/http_server.erl

```diff
@@ -3,73 +3,93 @@
 %%
 %% It will deliver the http-request params to initialer process
 %%--------------------------------------------------------------------
+%%
+%% Author:wwhai
+%%
 -module(http_server).
+-behaviour(gen_server).
 
--compile(export_all).
--compile(nowarn_export_all).
-
+-export([start_link/0]).
+-export([code_change/3, handle_call/3, handle_cast/2, handle_info/2, init/1, terminate/2]).
+-record(state, {}).
 %%--------------------------------------------------------------------
 %% APIs
 %%--------------------------------------------------------------------
-start() ->
-  {ok, _} = application:ensure_all_started(cowboy),
-  cowboy_router:compile([
-        {'_', [
-              {"/", ?MODULE, self()}
-        ]}
-  ]).
 
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+init([]) ->
+    EtsOptions = [named_table, public, set, {write_concurrency, true},
+                                            {read_concurrency, true}],
+    emqx_web_hook_http_test = ets:new(emqx_web_hook_http_test, EtsOptions),
+    start_http(),
+    start_https(),
+    {ok, #state{}}.
+
+handle_call(_Request, _From, State) ->
+    {reply, ignored, State}.
+
+handle_cast(_Msg, State) ->
+    {noreply, State}.
+
+handle_info(_Info, State) ->
+    {noreply, State}.
+
+terminate(_Reason, _State) ->
+    stop_http(),
+    stop_https().
+
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+%%--------------------------------------------------------------------
+%% Callbacks
+%%--------------------------------------------------------------------
 
 start_http() ->
```

Comment:
`start_http(Port) ->`

We'd better pass the port as an argument so that we can try another port it if binding failed.